lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -136,177 +145,262 @@ public void OnNext(IDriverStarted value)
         /// </summary>
         /// <param name="completedTask">The link to the completed task</param>
         public void OnNext(ICompletedTask completedTask)
         {
-            _completedTasks.Add(completedTask);
-
-            if (_completedTasks.Count != _dataSet.Count + 1)
+            lock (_completedTasksLock)
             {
-                return;
-            }
-            
-            foreach (var task in _completedTasks)
-            {
-                Logger.Log(Level.Verbose, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                Logger.Log(Level.Info,
+                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
+                _completedTasks.Add(completedTask);
+
+                if (_completedTasks.Count >= _totalMappers + 1)
+                {
+                    _areImruTasksCompleted = true;
+                    foreach (var task in _completedTasks)
+                    {
+                        Logger.Log(Level.Info, string.Format("Disposing task: {0}", task.Id));
+                        task.ActiveContext.Dispose();
+                    }
+                }
             }
         }
 
+        /// <summary>
+        /// Specifies what to do when evaluator fails.
+        /// If we get all completed tasks then ignore the failure
+        /// Else request a new evaluator. If failure happens in middle of IMRU 
+        /// job we expect neighboring evaluators to fail while doing 
+        /// communication and will use FailedTask and FailedContext logic to 
+        /// order shutdown.
+        /// </summary>
+        /// <param name="value"></param>
         public void OnNext(IFailedEvaluator value)
         {
-            Logger.Log(Level.Info, "An evaluator failed, checking if it failed before context and service was submitted");
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-
-            if (value.FailedContexts != null && value.FailedContexts.Count != 0)
+            if (_areImruTasksCompleted)
             {
-                Logger.Log(Level.Info, "Some active context failed, cannot continue IMRU task");        
-                Exceptions.Throw(new Exception(), Logger);
+                Logger.Log(Level.Info,
+                    string.Format("Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
+                return;
             }
 
+            Logger.Log(Level.Info,
+                string.Format("Evaluator with Id: {0} failed with Exception: {1}", value.Id, value.EvaluatorException));
+            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
             if (currFailedEvaluators > _allowedFailedEvaluators)
             {
                 Exceptions.Throw(new Exception("Cannot continue IMRU job, Failed evaluators reach maximum limit"),
```

Comment:
Add an exception class for this, e.g. `MaximumNumberOfEvalutorFailuresExceededException`. That way, we can report this more orderly to the client.